File: p114m/flatten_tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
const Tree = require('binary_tree');
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
var flatten = function (root) {
  let pre = null;
  // @param {TreeNode} current node
  // @return {TreeNode} the tail after flattened the child
  const helper = (p) => {
    if (p == null) return p;
    helper(p.right);
    helper(p.left);
    [p.left, p.right] = [null, pre];
    pre = p;
  };

  helper(root);
};

const f = (a, b) => [a, b];
[
  f('#', '#'),
  f('0,#,#', '0,#,#'),
  f('1,2,#,#,#', '1,#,2,#,#'),
  f('1,#,2,#,#', '1,#,2,#,#'),
  f('1,2,#,#,3,#,#', '1,#,2,#,3,#,#'),
  f('1,2,3,#,#,#,#', '1,#,2,#,3,#,#'),
  f('1,#,2,#,3,#,#', '1,#,2,#,3,#,#'),
  f('1,2,3,#,#,4,#,#,5,#,6,#,#', '1,#,2,#,3,#,4,#,5,#,6,#,#'),
].forEach(([tree, expected]) => {
  let root = Tree.deserialize(tree);
  flatten(root);
  let actual = Tree.serialize(root);
  console.log('Flatten', tree, '->', actual);
  console.assert(actual === expected);
});

File: p114m/flatten_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from local_packages.binary_tree import TreeNode


class Solution:
    def flattenV1(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """

        def helper(p: TreeNode) -> TreeNode:
            if not p or (not p.left and not p.right):
                return p
            right = p.right
            ltail = helper(p.left)
            if ltail:
                p.left, p.right = None, p.left
                ltail.right = right
            return helper(right) if right else ltail

        helper(root)

    def flattenV2(self, root: TreeNode) -> None:
        pre = None

        def helper(p: TreeNode) -> None:
            nonlocal pre
            if not p:
                return
            helper(p.right)
            helper(p.left)
            p.left, p.right = None, pre
            pre = p

        helper(root)


# TESTS
for tree, expected in [
    ("#", "#"),
    ("0,#,#", "0,#,#"),
    ("1,2,#,#,#", "1,#,2,#,#"),
    ("1,#,2,#,#", "1,#,2,#,#"),
    ("1,2,#,#,3,#,#", "1,#,2,#,3,#,#"),
    ("1,2,3,#,#,#,#", "1,#,2,#,3,#,#"),
    ("1,#,2,#,3,#,#", "1,#,2,#,3,#,#"),
    ("1,2,3,#,#,4,#,#,5,#,6,#,#", "1,#,2,#,3,#,4,#,5,#,6,#,#"),
]:
    sol = Solution()
    root = TreeNode.deserialize(tree)
    sol.flattenV2(root)
    actual = TreeNode.serialize(root)
    print("Flatten", tree, "->", actual)
    assert actual == expected
